service/resource/endpoint/create.go

```diff
@@ -1,11 +1,73 @@
 package endpoint
 
-import "context"
+import (
+	"context"
+
+	"github.com/giantswarm/microerror"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	apiv1 "k8s.io/client-go/pkg/api/v1"
+)
 
 func (r *Resource) GetCreateState(ctx context.Context, obj, currentState, desiredState interface{}) (interface{}, error) {
-	return nil, nil
+	currentEndpoint, err := toEndpoint(currentState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	desiredEndpoint, err := toEndpoint(desiredState)
+	if err != nil {
+		return nil, microerror.Mask(err)
+	}
+
+	createState := Endpoint{
+		ServiceName:      currentEndpoint.ServiceName,
+		ServiceNamespace: currentEndpoint.ServiceNamespace,
+	}
+	for _, currentIP := range currentEndpoint.IPs {
+		if !containsIP(createState.IPs, currentIP) {
+			createState.IPs = append(createState.IPs, currentIP)
+		}
+	}
+	for _, desiredIP := range desiredEndpoint.IPs {
+		if !containsIP(createState.IPs, desiredIP) {
+			createState.IPs = append(createState.IPs, desiredIP)
+		}
+	}
+
+	return createState, nil
 }
 
 func (r *Resource) ProcessCreateState(ctx context.Context, obj, createState interface{}) error {
+	endpointToCreate, err := toEndpoint(createState)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	if len(endpointToCreate.IPs) == 0 {
+		// Nothing to do.
+		return nil
+	}
+
+	k8sAddresses := []apiv1.EndpointAddress{}
+	for _, endpointIP := range endpointToCreate.IPs {
+		k8sAddress := apiv1.EndpointAddress{
+			IP: endpointIP,
+		}
+		k8sAddresses = append(k8sAddresses, k8sAddress)
+	}
+
+	k8sEndpoint, err := r.k8sClient.CoreV1().Endpoints(endpointToCreate.ServiceNamespace).Get(endpointToCreate.ServiceName, metav1.GetOptions{})
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	for i := range k8sEndpoint.Subsets {
+		k8sEndpoint.Subsets[i].Addresses = k8sAddresses
+	}
```

Comment:
`ProcessCreateState` looks way better now than the former version. Anyway, I think everything starting at line 51 up to here should not be done here. The process methods should only take state and update it against some external dependency, if any. Here is still state computation that has to be done to prepare the actual update. This preparation should all be part of `GetCreateState` in this case. 

A learning we can take from here is probably that the state that is actually used to update against some external dependency is actually the state that should be dispatched between the resource steps. In this case this means the resource and thus the state which is reconciled is the Kubernetes endpoint. So `GetCurrentState` should fetch the endpoint as it is. `GetDesiredState` should compute the endpoint as it should be. This then includes all the magic we do here and above. In this case it also makes sense to me to rather create a composition between current and desired state, as opposed to creating a diff than usual. The system and the resource just require us to do this. `GetCreateState` should be super dumb and return the desired state in case it does not match the current state. `ProcessCreateState` should only take the create state and call the kubernetes client's update method, not more.

Reply:
This would mean that the `createState` is a k8s Endpoint so `apiv1.Endpoint` instead of using the struct defined in `spec`, or did I misunderstand you?

I will create an issue to take care of this after the basic testing is done.

Reply:
You understood right.